src/Middleware/Spa/SpaProxy/src/SpaProxyLaunchManager.cs

```diff
@@ -0,0 +1,213 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+using Microsoft.Extensions.Logging;
+using System;
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.AspNetCore.SpaProxy
+{
+    internal class SpaProxyLaunchManager : IHostedService, IDisposable
+    {
+        private readonly SpaDevelopmentServerOptions _options;
+        private readonly ILogger<SpaProxyLaunchManager> _logger;
+        private readonly HttpClient _httpClient = new(new HttpClientHandler()
+        {
+            // It's ok for us to do this here since this service is only plugged in during development.
+            ServerCertificateCustomValidationCallback = HttpClientHandler.DangerousAcceptAnyServerCertificateValidator
+        });
+
+        private Process? _spaProcess;
+        private bool _disposedValue;
+
+        public SpaProxyLaunchManager(ILogger<SpaProxyLaunchManager> logger)
+        {
+            _options = new SpaDevelopmentServerOptions();
+            var configuration = new ConfigurationBuilder()
+                .AddJsonFile(Path.Combine(AppContext.BaseDirectory, "spa.proxy.json"))
+                .Build();
+            configuration.GetSection("SpaProxyServer").Bind(_options);
+            _logger = logger;
+        }
+
+        public async Task StartAsync(CancellationToken cancellationToken)
+        {
+            _logger.LogInformation("Starting SPA development server");
+            var running = await ProbeSpaDevelopmentServerUrl(cancellationToken);
+            if (running)
+            {
+                _logger.LogInformation($"Found SPA development server running at {_options.ServerUrl}");
+            }
+            else
+            {
+                _logger.LogInformation($"No SPA development server running at {_options.ServerUrl} found.");
+                await StartSpaProcessAndProbeForLiveness(cancellationToken);
+            }
+        }
+
+        private async Task<bool> ProbeSpaDevelopmentServerUrl(CancellationToken cancellationToken)
+        {
+            try
+            {
+                var response = await _httpClient.GetAsync(_options.ServerUrl, cancellationToken);
```

Comment:
How long might this call last? Do we generally find that the SPA dev servers reply instantly if they are alive, or do they potentially wait minutes while going through a webpack build?

Maybe even if they do take ages it's desirable to wait for them. Just asking to clear up my understanding of the startup flow.

Reply:
It comes back right away. However I've updated the code to add a timeout after 1 second or if the host request cancellation, whatever happens first.